kehadiran_scraper/main.py

```diff
@@ -1,93 +1,158 @@
+        cur_date: dt.date = max_date - dt.timedelta(days=i)
+        if use_whitelist:
+            if cur_date not in whitelist:
+                continue
         url: str = URL_PATTERN.format(
             day=str(cur_date.day).zfill(2),
             month=str(cur_date.month).zfill(2),
             year=cur_date.year,
         )
-        yield url
+        yield url, cur_date
+
+
+async def _get_whitelist_dates_from_file() -> set[dt.date]:
+    s: set[dt.date] = set()
+    async with aiofiles.open(WHITE_LIST_FILENAME, mode="r") as fi:
+        async for line in fi:
+            date = dt.date.fromisoformat(line.strip("\n"))
+            s.add(date)
+    return s
+
+
+async def _update_whitelist_to_file(whitelist: set[dt.date]) -> None:
+    async with aiofiles.open(WHITE_LIST_FILENAME, mode="w") as fo:
+        whitelist_sorted = "\n".join(sorted(str(d) for d in whitelist))
+        await fo.write(whitelist_sorted)
 
 
 async def download(
     url: str,
     session: aiohttp.ClientSession,
 ) -> tuple[t.Optional[bytes], t.Optional[str]]:
-    logger.info("Downloading from %s", url)
+    logger.debug("Downloading from %s", url)
     async with session.get(url) as response:
         data: bytes = await response.read()
         if b"File does not exist" in data:
-            # TODO: Collect url in a blacklist set. Later we can dump into file
-            logger.error("Url %s does not contain file. Need to blacklist", url)
-            return (None, None)
+            logger.debug("Url %s does not contain file", url)
+            return (None, url)
         return data, url
 
 
-async def save(data: bytes, url: str):
-    # TODO: Write to S3
-    filename: str = TEST_OUTPUT_DIR + url.split("/")[-1]
+async def save(data: bytes, url: str, date: dt.date):
+    filename: str = OUTPUT_DIR + dt.date.strftime(date, "%Y-%m-%d") + ".pdf"
     logger.info("Found file from %s, dumping to file %s", url, filename)
     async with aiofiles.open(filename, mode="wb") as fo:
         await fo.write(data)
 
 
-async def scrape():
+async def scrape(min_date: dt.date, max_date: dt.date, visited_dates: list[dt.date]):
+    whitelist: set[dt.date] = await _get_whitelist_dates_from_file()
+
     async with aiohttp.ClientSession() as session:
-        # TODO: Skip url found in blacklist file (Or use whitelist instead)
-        for urls in _chunker(_url_generator(start=FIRST_DATE), size=SCRAPE_CHUNK_SIZE):
-            download_tasks = [download(url=url, session=session) for url in urls]
+        urls_iter: t.Iterator[tuple[str, dt.date]] = _url_generator(
+            min_date=min_date,
+            max_date=max_date,
+            whitelist=whitelist,
+            use_whitelist=USE_WHITELIST,
+        )
+
+        for urls in _chunker(urls_iter, size=SCRAPE_CHUNK_SIZE):
+            url_to_date: dict[str, dt.date] = {}
+            download_tasks = []
+            for url, cur_date in urls:
+                task = download(url=url, session=session)
```

Comment:
I think we should stagger the download with a random delay so the receiving server doesn't get all the requests at the same time. This would help to load balance the request & avoid getting banned by them.